File: android/app/src/main/java/org/chamilo/mobile/ChamiloSecureStoragePlugin.java
package org.chamilo.mobile;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.io.IOException;
import java.security.GeneralSecurityException;
import org.json.JSONObject;

@CapacitorPlugin(name = "ChamiloSecureStorage")
public class ChamiloSecureStoragePlugin extends Plugin {

    private static final int MAX_KEY_LENGTH = 512;
    private static final int MAX_VALUE_LENGTH = 64 * 1024;

    private SecureStorageVault vault;

    @PluginMethod
    public void get(PluginCall call) {
        String key = validKey(call);

        if (key == null) {
            return;
        }

        try {
            String value = vault().read(key);
            JSObject result = new JSObject();
            result.put("value", value == null ? JSONObject.NULL : value);
            call.resolve(result);
        } catch (IOException | GeneralSecurityException error) {
            call.reject("Secure storage read failed.");
        }
    }

    @PluginMethod
    public void set(PluginCall call) {
        String key = validKey(call);
        String value = call.getString("value");

        if (key == null) {
            return;
        }

        if (value == null || value.length() > MAX_VALUE_LENGTH) {
            call.reject("Secure storage value is invalid.");
            return;
        }

        try {
            vault().write(key, value);
            call.resolve();
        } catch (IOException | GeneralSecurityException error) {
            call.reject("Secure storage write failed.");
        }
    }

    @PluginMethod
    public void remove(PluginCall call) {
        String key = validKey(call);

        if (key == null) {
            return;
        }

        try {
            vault().remove(key);
            call.resolve();
        } catch (IOException error) {
            call.reject("Secure storage removal failed.");
        }
    }

    private String validKey(PluginCall call) {
        String key = call.getString("key");

        if (key == null || key.trim().isEmpty() || key.length() > MAX_KEY_LENGTH) {
            call.reject("Secure storage key is invalid.");
            return null;
        }

        return key;
    }

    private SecureStorageVault vault() {
        if (vault == null) {
            vault = new SecureStorageVault(getContext());
        }

        return vault;
    }
}

File: ios/App/App/ChamiloSecureStoragePlugin.swift
import Capacitor
import Foundation
import Security

private enum ChamiloKeychainError: Error {
    case unexpectedStatus(OSStatus)
    case invalidStoredValue
}

private final class ChamiloKeychainStore {
    private let service = "org.chamilo.mobile.secure-storage.v1"

    func read(key: String) throws -> String? {
        var query = baseQuery(key: key)
        query[kSecReturnData] = true
        query[kSecMatchLimit] = kSecMatchLimitOne

        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)

        if status == errSecItemNotFound {
            return nil
        }

        guard status == errSecSuccess else {
            throw ChamiloKeychainError.unexpectedStatus(status)
        }

        guard
            let data = item as? Data,
            let value = String(data: data, encoding: .utf8)
        else {
            try? remove(key: key)
            throw ChamiloKeychainError.invalidStoredValue
        }

        return value
    }

    func write(key: String, value: String) throws {
        let valueData = Data(value.utf8)
        let query = baseQuery(key: key)
        let updateAttributes: [CFString: Any] = [
            kSecValueData: valueData
        ]

        let updateStatus = SecItemUpdate(
            query as CFDictionary,
            updateAttributes as CFDictionary
        )

        if updateStatus == errSecSuccess {
            return
        }

        guard updateStatus == errSecItemNotFound else {
            throw ChamiloKeychainError.unexpectedStatus(updateStatus)
        }

        var newItem = query
        newItem[kSecValueData] = valueData
        newItem[kSecAttrAccessible] = kSecAttrAccessibleWhenUnlockedThisDeviceOnly

        let addStatus = SecItemAdd(newItem as CFDictionary, nil)

        guard addStatus == errSecSuccess else {
            throw ChamiloKeychainError.unexpectedStatus(addStatus)
        }
    }

    func remove(key: String) throws {
        let status = SecItemDelete(baseQuery(key: key) as CFDictionary)

        guard status == errSecSuccess || status == errSecItemNotFound else {
            throw ChamiloKeychainError.unexpectedStatus(status)
        }
    }

    private func baseQuery(key: String) -> [CFString: Any] {
        [
            kSecClass: kSecClassGenericPassword,
            kSecAttrService: service,
            kSecAttrAccount: key
        ]
    }
}

@objc(ChamiloSecureStoragePlugin)
public class ChamiloSecureStoragePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ChamiloSecureStoragePlugin"
    public let jsName = "ChamiloSecureStorage"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "get", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "set", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "remove", returnType: CAPPluginReturnPromise)
    ]

    private static let maxKeyLength = 512
    private static let maxValueLength = 64 * 1024
    private let store = ChamiloKeychainStore()

    @objc func get(_ call: CAPPluginCall) {
        guard let key = validKey(call) else {
            return
        }

        do {
            let value = try store.read(key: key)

            if let value {
                call.resolve(["value": value])
            } else {
                call.resolve(["value": NSNull()])
            }
        } catch {
            call.reject("Secure storage read failed.")
        }
    }

    @objc func set(_ call: CAPPluginCall) {
        guard let key = validKey(call) else {
            return
        }

        guard
            let value = call.getString("value"),
            value.utf16.count <= Self.maxValueLength
        else {
            call.reject("Secure storage value is invalid.")
            return
        }

        do {
            try store.write(key: key, value: value)
            call.resolve()
        } catch {
            call.reject("Secure storage write failed.")
        }
    }

    @objc func remove(_ call: CAPPluginCall) {
        guard let key = validKey(call) else {
            return
        }

        do {
            try store.remove(key: key)
            call.resolve()
        } catch {
            call.reject("Secure storage removal failed.")
        }
    }

    private func validKey(_ call: CAPPluginCall) -> String? {
        guard
            let key = call.getString("key"),
            !key.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
            key.utf16.count <= Self.maxKeyLength
        else {
            call.reject("Secure storage key is invalid.")
            return nil
        }

        return key
    }
}
